### ISSUE TITLE
install mxnet under centos

### ISSUE THREAD

**Issue by JulySouthWind** [dmlc/mxnet#1172]
I  am installing mxnet under centos, I builded opencv and openblas from source,but these error happened when run make -j4.
Anyone can help me? Thanks
/usr/local/bin/g++ -DMSHADOW_FORCE_STREAM -Wall -O3 -I./mshadow/ -I./dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CUDA=0 -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=1 -I/data2/opencv/opencv-2.4/build/include  -fopenmp -I/usr/local/openblas/include -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/resource.o build/c_api/c_api.o build/c_api/c_api_error.o build/c_api/c_predict_api.o build/common/mxrtc.o build/common/tblob_op_registry.o build/engine/engine.o build/engine/naive_engine.o build/engine/threaded_engine.o build/engine/threaded_engine_perdevice.o build/engine/threaded_engine_pooled.o build/io/io.o build/io/iter_csv.o build/io/iter_image_recordio.o build/io/iter_mnist.o build/kvstore/kvstore.o build/ndarray/ndarray.o build/ndarray/ndarray_function.o build/ndarray/unary_function.o build/operator/activation.o build/operator/batch_norm.o build/operator/block_grad.o build/operator/concat.o build/operator/convolution.o build/operator/crop.o build/operator/cross_device_copy.o build/operator/cudnn_batch_norm.o build/operator/deconvolution.o build/operator/dropout.o build/operator/elementwise_binary_op.o build/operator/elementwise_binary_scalar_op.o build/operator/elementwise_sum.o build/operator/embedding.o build/operator/fully_connected.o build/operator/identity_attach_KL_sparse_reg.o build/operator/leaky_relu.o build/operator/lrn.o build/operator/native_op.o build/operator/ndarray_op.o build/operator/operator.o build/operator/pooling.o build/operator/regression_output.o build/operator/reshape.o build/operator/slice_channel.o build/operator/softmax_activation.o build/operator/softmax_output.o build/operator/swapaxis.o build/operator/upsampling.o build/optimizer/optimizer.o build/optimizer/sgd.o build/storage/storage.o build/symbol/graph_executor.o build/symbol/static_graph.o build/symbol/symbol.o dmlc-core/libdmlc.a -pthread -lm -lopenblas -lrt -L/data2/opencv/opencv-2.4/build/lib -L/usr/local/openblas/lib
/tmp/ccd27Zq4.o: In function `cv::Mat::~Mat()':
im2rec.cc:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x79): undefined reference to `cv::Mat::deallocate()'
im2rec.cc:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x69): undefined reference to `cv::fastFree(void*)'
/tmp/ccd27Zq4.o: In function `main':
im2rec.cc:(.text.startup+0x1e54): undefined reference to `vtable for cv::_InputArray'
im2rec.cc:(.text.startup+0x1ea0): undefined reference to `cv::imdecode(cv::_InputArray const&, int)'
im2rec.cc:(.text.startup+0x200d): undefined reference to `cv::Mat::Mat(cv::Mat const&, cv::Range const&, cv::Range const&)'
im2rec.cc:(.text.startup+0x21ea): undefined reference to `cv::fastFree(void*)'
im2rec.cc:(.text.startup+0x225d): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
im2rec.cc:(.text.startup+0x2270): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
im2rec.cc:(.text.startup+0x2290): undefined reference to `cv::resize(cv::_InputArray const&, cv::_OutputArray const&, cv::Size_<int>, double, double, int)'
im2rec.cc:(.text.startup+0x22b0): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
im2rec.cc:(.text.startup+0x22cc): undefined reference to `cv::imencode(std::string const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
im2rec.cc:(.text.startup+0x23ca): undefined reference to `cv::Mat::deallocate()'
im2rec.cc:(.text.startup+0x243c): undefined reference to `cv::fastFree(void*)'
im2rec.cc:(.text.startup+0x2460): undefined reference to `cv::Mat::deallocate()'
im2rec.cc:(.text.startup+0x24d5): undefined reference to `cv::fastFree(void*)'
im2rec.cc:(.text.startup+0x2c6a): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
im2rec.cc:(.text.startup+0x2cc6): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
im2rec.cc:(.text.startup+0x2cd9): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
im2rec.cc:(.text.startup+0x2cf9): undefined reference to `cv::resize(cv::_InputArray const&, cv::_OutputArray const&, cv::Size_<int>, double, double, int)'
im2rec.cc:(.text.startup+0x2deb): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
im2rec.cc:(.text.startup+0x2df6): undefined reference to `cv::Mat::copyTo(cv::_OutputArray const&) const'
im2rec.cc:(.text.startup+0x2f5a): undefined reference to `cv::Mat::deallocate()'
im2rec.cc:(.text.startup+0x2fd3): undefined reference to `cv::fastFree(void*)'
……
build/io/iter_image_recordio.o: In function `mxnet::io::ImageRecordIOParser::Init(std::vector<std::pair<std::string, std::string>, std::allocator<std::pair<std::string, std::string> > > const&)':
iter_image_recordio.cc:(.text._ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE[_ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE]+0x334): undefined reference to `cv::Mat::create(int, int const*, int)'
iter_image_recordio.cc:(.text._ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE[_ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE]+0x3de): undefined reference to `cv::fastFree(void*)'
iter_image_recordio.cc:(.text._ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE[_ZN5mxnet2io19ImageRecordIOParser4InitERKSt6vectorISt4pairISsSsESaIS4_EE]+0x6c9): undefined reference to `cv::Mat::deallocate()'
collect2: error: ld returned 1 exit status
make: *** [bin/im2rec] Error 1

**Comment by thirdwing** [comment]
It seems that you only add `-L/data2/opencv/opencv-2.4/build/lib` in the makefile.

Personally I think this is not enough.

**Comment by JulySouthWind** [comment]
@thirdwing I followed this issue https://github.com/dmlc/mxnet/issues/331 .Do you have any suggestion?

**Comment by qiaohaijun** [comment]
I am running mxnet on centos 7.1.

but I install opencv and openblas through yum.

**Comment by JulySouthWind** [comment]
Thank you.  I tried，but not succeed. Maybe the conf of yum isn't fine

**Comment by leoli2012** [comment]
build mxnet under centos have a little complex, i spend several days to resolve this issue...however, i success last....
my os is centos 5.8 ,run on VM.
cd /home
myPath='install_mxnet'
if [ -x "$myPath" ]; then 
rm -rf "$myPath" 
fi 
mkdir install_mxnet
cd install_mxnet
yum -y install git
if [ "$?" == "0"]; then
  echo "git install fail"
  exit 0
fi

######update gcc to 4.8.2,if you gcc version <4.8,you must update gcc first 
new_gcc='/opt/rh/devtoolset-2/root/usr/bin/gcc'
cd /etc/yum.repos.d
wget http://people.centos.org/tru/devtools-2/devtools-2.repo
yum -y install devtoolset-2-gcc devtoolset-2-binutils devtoolset-2-gcc-c++
if [ ! -f "$new_gcc" ]; then
  echo "gcc install fail"
  exit 0
fi
###centos 5.8 gcc version is 4.1.2, you should rename gcc.bakxxx, which depend  your gcc version
mv -f /usr/bin/gcc /usr/bin/gcc.bak4.1.2
mv -f /usr/bin/c++ /usr/bin/c++.bak4.1.2
mv -f /usr/bin/g++ /usr/bin/g++.bak4.1.2

ln -s /opt/rh/devtoolset-2/root/usr/bin/gcc /usr/bin/gcc
ln -s /opt/rh/devtoolset-2/root/usr/bin/c++ /usr/bin/c++
ln -s /opt/rh/devtoolset-2/root/usr/bin/g++ /usr/bin/g++
cd -


######atlas
libcblas='/usr/lib64/atlas/libcblas.so'
rpm -ivh http://mirrors.yun-idc.com/epel/5Server/x86_64/epel-release-5-4.noarch.rpm
yum -y install atlas-devel
if [ ! -f "$libcblas" ]; then
  echo "atlas install fail"
  exit 0
fi
ln -s /usr/lib64/atlas/libcblas.so /usr/lib/libcblas.so

######update python, if your python version <2.7, update python to 2.7 first
new_python='/usr/local/bin/python2.7'
yum -y groupinstall 'Development Tools'
yum -y install openssl-devel* ncurses-devel* zlib*.x86_64
yum -y install bzip2 bzip2-devel bzip2-libs

wget https://www.python.org/ftp/python/2.7.9/Python-2.7.9.tgz
tar -xzf Python-2.7.9.tgz
cd Python-2.7.9
./configure
make
make install
if [ ! -f "$new_python" ]; then
  echo "python update fail"
  exit 0
fi
mv /usr/bin/python /usr/bin/python2.4.3
ln -s /usr/local/bin/python2.7 /usr/bin/python
sed -i 's/#!\/usr\/bin\/python/#!\/usr\/bin\/python2.4.3/g'  /usr/bin/yum 
cd -

#######install setuptools
wget --no-check-certificate https://bootstrap.pypa.io/ez_setup.py
python ez_setup.py --insecure

######install numpy
wget http://sourceforge.net/projects/numpy/files/NumPy/1.10.2/numpy-1.10.2.tar.gz
if [ "$?" == "0" ];then
  echo "numpy download fail"
  exit 0
fi
tar -xzf numpy-1.10.2.tar.gz 
cd numpy-1.10.2
python setup.py build
python setup.py install
cd -

######install mxnet
git clone --recursive https://github.com/dmlc/mxnet
if [ "$?" == "0" ]; then
  echo "mxnet download fail"
  exit 0
fi
cd mxnet
cp make/config.mk .
sed -i 's/USE_OPENCV = 1/USE_OPENCV = 0/g' config.mk     
sed -i 's/USE_DIST_KVSTORE = 0/USE_DIST_KVSTORE = 1/g' config.mk
sed -i 's/wget/wget\ --no-check-certificate/g' ps-lite/make/deps.mk
make -j4
-----------------------------------------
have a try ~

**Comment by thirdwing** [comment]
@JulySouthWind By adding `-L/data2/opencv/opencv-2.4/build/lib`, you tell compiler where to find the opencv libraries. But you also need to tell compiler which libraries to link, like `-lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d`
